File: labs/control-ui/api-shadow/adaptive_strategy_racer.ts
/**
 * Adaptive Strategy Racer & Disposable Fake SNI Probe Engine
 * Originates from UAC-SNI-Spoofer-Windows-main and adapted for LumiNet Control UI.
 */

import { locateSni, fixedChunks, splitAt, splitTlsRecord } from './multi_strategy_fragmenter';

export type ExtendedFragmentStrategy =
  | 'SNI_CHARS'
  | 'MULTI64'
  | 'FULL5'
  | 'FULL10'
  | 'FULL20'
  | 'SNI_BOUNDARY'
  | 'SNI_SPLIT'
  | 'TLS_RECORD_FRAG'
  | 'TLS_SNI_RECORDS'
  | 'HALF'
  | 'RAW';

export type StrategyResponseStatus =
  'VALID' | 'EMPTY_RESPONSE' | 'ALERT_REJECTED' | 'TRUNCATED_MALFORMED';

export interface StrategyPlanEntry {
  strategy: ExtendedFragmentStrategy;
  delayMs: number;
}

export type CarrierMode = 'MCI' | 'IRANCELL' | 'ADAPTIVE';

export function validateStrategyResponse(
  response: Uint8Array | null | undefined,
): StrategyResponseStatus {
  if (!response || response.length === 0) {
    return 'EMPTY_RESPONSE';
  }
  const first = response[0];
  if (first === 0x15) {
    return 'ALERT_REJECTED';
  }
  if (response.length < 8 && (first === 0x14 || first === 0x16 || first === 0x17)) {
    return 'TRUNCATED_MALFORMED';
  }
  return 'VALID';
}

export function fragmentExtended(
  data: Uint8Array,
  strategy: ExtendedFragmentStrategy,
  chunkSize: number = 64,
): Uint8Array[] {
  if (data.length < 2) return [data.slice()];
  const location = locateSni(data);

  switch (strategy) {
    case 'RAW':
      return [data.slice()];
    case 'HALF':
      return splitAt(data, Math.floor(data.length / 2));
    case 'FULL5':
      return fixedChunks(data, 5);
    case 'FULL10':
      return fixedChunks(data, 10);
    case 'FULL20':
      return fixedChunks(data, 20);
    case 'MULTI64':
      return fixedChunks(data, chunkSize > 0 ? chunkSize : 64);
    case 'SNI_BOUNDARY':
      return location ? splitAt(data, location[0]) : splitAt(data, Math.floor(data.length / 2));
    case 'SNI_SPLIT':
      return location
        ? splitAt(data, location[0] + Math.max(1, Math.floor(location[1] / 2)))
        : splitAt(data, Math.floor(data.length / 2));
    case 'SNI_CHARS': {
      if (!location) {
        return splitAt(data, Math.floor(data.length / 2));
      }
      const [start, length] = location;
      const out: Uint8Array[] = [];
      if (start > 0) {
        out.push(data.slice(0, start));
      }
      for (let i = 0; i < length; i++) {
        out.push(data.slice(start + i, start + i + 1));
      }
      if (start + length < data.length) {
        out.push(data.slice(start + length));
      }
      return out;
    }
    case 'TLS_RECORD_FRAG':
      return splitTlsRecord(data, false);
    case 'TLS_SNI_RECORDS':
      return splitTlsRecord(data, true);
    default:
      return [data.slice()];
  }
}

export function buildDisposableFakeProbe(fakeSni: string): Uint8Array {
  const hostBytes = new TextEncoder().encode(fakeSni);
  const pkt: number[] = [];

  // Record Header
  pkt.push(0x16, 0x03, 0x01, 0x00, 0x00);

  const handshakeStart = pkt.length;
  pkt.push(0x01, 0x00, 0x00, 0x00); // ClientHello
  pkt.push(0x03, 0x03); // TLS 1.2

  for (let i = 0; i < 32; i++) {
    pkt.push((i * 37 + 11) & 0xff);
  }
  pkt.push(0x00); // Session ID len 0

  // Cipher Suites
  pkt.push(0x00, 0x04, 0x13, 0x01, 0x13, 0x03);
  pkt.push(0x01, 0x00); // Compression null

  const extLenPos = pkt.length;
  pkt.push(0x00, 0x00);
  const extStart = pkt.length;

  // SNI Extension
  const sniExtLen = 2 + 1 + 2 + hostBytes.length;
  pkt.push(0x00, 0x00, (sniExtLen >> 8) & 0xff, sniExtLen & 0xff);
  const listLen = 1 + 2 + hostBytes.length;
  pkt.push(
    (listLen >> 8) & 0xff,
    listLen & 0xff,
    0x00,
    (hostBytes.length >> 8) & 0xff,
    hostBytes.length & 0xff,
  );
  pkt.push(...hostBytes);

  // ALPN Extension (http/1.1, h2)
  pkt.push(
    0x00,
    0x10,
    0x00,
    0x0e,
    0x00,
    0x0c,
    0x08,
    0x68,
    0x74,
    0x74,
    0x70,
    0x2f,
    0x31,
    0x2e,
    0x31, // http/1.1
    0x02,
    0x68,
    0x32, // h2
  );

  const extLen = pkt.length - extStart;
  pkt[extLenPos] = (extLen >> 8) & 0xff;
  pkt[extLenPos + 1] = extLen & 0xff;

  const hsLen = pkt.length - handshakeStart - 4;
  pkt[handshakeStart + 1] = (hsLen >> 16) & 0xff;
  pkt[handshakeStart + 2] = (hsLen >> 8) & 0xff;
  pkt[handshakeStart + 3] = hsLen & 0xff;

  const recLen = pkt.length - 5;
  pkt[3] = (recLen >> 8) & 0xff;
  pkt[4] = recLen & 0xff;

  return new Uint8Array(pkt);
}

export class AdaptiveStrategyRacer {
  private preferredStrategies: Map<string, ExtendedFragmentStrategy> = new Map();
  public carrierMode: CarrierMode;
  public fakeSni: string;
  public fakeProbeEnabled: boolean;
  constructor(
    carrierMode: CarrierMode = 'ADAPTIVE',
    fakeSni: string = 'www.speedtest.net',
    fakeProbeEnabled: boolean = true,
  ) {
    this.carrierMode = carrierMode;
    this.fakeSni = fakeSni;
    this.fakeProbeEnabled = fakeProbeEnabled;
  }

  planStrategies(host: string): StrategyPlanEntry[] {
    let base: StrategyPlanEntry[] = [];
    switch (this.carrierMode) {
      case 'MCI':
        base = [
          { strategy: 'FULL20', delayMs: 1 },
          { strategy: 'FULL10', delayMs: 2 },
          { strategy: 'FULL5', delayMs: 5 },
          { strategy: 'SNI_CHARS', delayMs: 2 },
          { strategy: 'SNI_BOUNDARY', delayMs: 1 },
          { strategy: 'SNI_SPLIT', delayMs: 3 },
          { strategy: 'TLS_RECORD_FRAG', delayMs: 3 },
          { strategy: 'TLS_SNI_RECORDS', delayMs: 3 },
          { strategy: 'HALF', delayMs: 3 },
          { strategy: 'RAW', delayMs: 0 },
        ];
        break;
      case 'IRANCELL':
        base = [
          { strategy: 'MULTI64', delayMs: 0 },
          { strategy: 'SNI_BOUNDARY', delayMs: 0 },
          { strategy: 'TLS_RECORD_FRAG', delayMs: 1 },
          { strategy: 'SNI_CHARS', delayMs: 1 },
          { strategy: 'SNI_SPLIT', delayMs: 2 },
          { strategy: 'HALF', delayMs: 2 },
          { strategy: 'RAW', delayMs: 0 },
        ];
        break;
      case 'ADAPTIVE':
      default:
        base = [
          { strategy: 'SNI_SPLIT', delayMs: 2 },
          { strategy: 'SNI_BOUNDARY', delayMs: 1 },
          { strategy: 'SNI_CHARS', delayMs: 1 },
          { strategy: 'TLS_SNI_RECORDS', delayMs: 2 },
          { strategy: 'TLS_RECORD_FRAG', delayMs: 2 },
          { strategy: 'MULTI64', delayMs: 0 },
          { strategy: 'HALF', delayMs: 2 },
          { strategy: 'RAW', delayMs: 0 },
        ];
        break;
    }

    const preferred = this.preferredStrategies.get(host);
    if (!preferred) return base;

    const prioritized: StrategyPlanEntry[] = [];
    const others: StrategyPlanEntry[] = [];
    for (const entry of base) {
      if (entry.strategy === preferred) {
        prioritized.push(entry);
      } else {
        others.push(entry);
      }
    }
    return [...prioritized, ...others];
  }

  recordSuccess(host: string, strategy: ExtendedFragmentStrategy): void {
    this.preferredStrategies.set(host, strategy);
  }

  preferredStrategy(host: string): ExtendedFragmentStrategy | undefined {
    return this.preferredStrategies.get(host);
  }

  buildFakeProbe(): Uint8Array | null {
    if (!this.fakeProbeEnabled || !this.fakeSni) return null;
    return buildDisposableFakeProbe(this.fakeSni);
  }
}

File: labs/mobile/donor-archive/tunnel/AdaptiveStrategyRacer.kt
package com.luminet.android.tunnel

import java.nio.charset.StandardCharsets

/**
 * Granular fragmentation strategies including byte-per-character SNI splitting.
 */
enum class ExtendedFragmentStrategy(val rawName: String) {
    SNI_CHARS("sni_chars"),
    MULTI64("multi64"),
    FULL5("full5"),
    FULL10("full10"),
    FULL20("full20"),
    SNI_BOUNDARY("sni_boundary"),
    SNI_SPLIT("sni_split"),
    TLS_RECORD_FRAG("tls_record_frag"),
    TLS_SNI_RECORDS("tls_sni_records"),
    HALF("half"),
    RAW("raw"),
}

enum class StrategyResponseStatus {
    VALID,
    EMPTY_RESPONSE,
    ALERT_REJECTED,
    TRUNCATED_MALFORMED,
}

data class StrategyPlanEntry(
    val strategy: ExtendedFragmentStrategy,
    val delayMs: Int,
)

enum class CarrierMode {
    MCI,
    IRANCELL,
    ADAPTIVE,
}

class AdaptiveStrategyRacer(
    private val carrierMode: CarrierMode = CarrierMode.ADAPTIVE,
    private val fakeSni: String = "www.speedtest.net",
    private val fakeProbeEnabled: Boolean = true,
) {
    private val preferredStrategies = mutableMapOf<String, ExtendedFragmentStrategy>()

    @Synchronized
    fun planStrategies(host: String): List<StrategyPlanEntry> {
        val base = when (carrierMode) {
            CarrierMode.MCI -> listOf(
                StrategyPlanEntry(ExtendedFragmentStrategy.FULL20, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.FULL10, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.FULL5, 5),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_CHARS, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_BOUNDARY, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_SPLIT, 3),
                StrategyPlanEntry(ExtendedFragmentStrategy.TLS_RECORD_FRAG, 3),
                StrategyPlanEntry(ExtendedFragmentStrategy.TLS_SNI_RECORDS, 3),
                StrategyPlanEntry(ExtendedFragmentStrategy.HALF, 3),
                StrategyPlanEntry(ExtendedFragmentStrategy.RAW, 0),
            )
            CarrierMode.IRANCELL -> listOf(
                StrategyPlanEntry(ExtendedFragmentStrategy.MULTI64, 0),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_BOUNDARY, 0),
                StrategyPlanEntry(ExtendedFragmentStrategy.TLS_RECORD_FRAG, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_CHARS, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_SPLIT, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.HALF, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.RAW, 0),
            )
            CarrierMode.ADAPTIVE -> listOf(
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_SPLIT, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_BOUNDARY, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.SNI_CHARS, 1),
                StrategyPlanEntry(ExtendedFragmentStrategy.TLS_SNI_RECORDS, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.TLS_RECORD_FRAG, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.MULTI64, 0),
                StrategyPlanEntry(ExtendedFragmentStrategy.HALF, 2),
                StrategyPlanEntry(ExtendedFragmentStrategy.RAW, 0),
            )
        }

        val preferred = preferredStrategies[host] ?: return base
        val (first, rest) = base.partition { it.strategy == preferred }
        return first + rest
    }

    @Synchronized
    fun recordSuccess(host: String, strategy: ExtendedFragmentStrategy) {
        preferredStrategies[host] = strategy
    }

    @Synchronized
    fun preferredStrategy(host: String): ExtendedFragmentStrategy? {
        return preferredStrategies[host]
    }

    fun buildFakeProbe(): ByteArray? {
        if (!fakeProbeEnabled || fakeSni.isEmpty()) return null
        return buildDisposableFakeProbe(fakeSni)
    }

    companion object {
        fun validateStrategyResponse(response: ByteArray?): StrategyResponseStatus {
            if (response == null || response.isEmpty()) return StrategyResponseStatus.EMPTY_RESPONSE
            val first = response[0].toInt() and 0xff
            if (first == 0x15) return StrategyResponseStatus.ALERT_REJECTED
            if (response.size < 8 && (first == 0x14 || first == 0x16 || first == 0x17)) {
                return StrategyResponseStatus.TRUNCATED_MALFORMED
            }
            return StrategyResponseStatus.VALID
        }

        fun fragmentExtended(
            data: ByteArray,
            strategy: ExtendedFragmentStrategy,
            chunkSize: Int = 64,
        ): List<ByteArray> {
            if (data.size < 2) return listOf(data.copyOf())
            val location = MultiStrategyFragmenter.locateSni(data)

            return when (strategy) {
                ExtendedFragmentStrategy.RAW -> listOf(data.copyOf())
                ExtendedFragmentStrategy.HALF -> {
                    val cut = maxOf(1, data.size / 2)
                    listOf(data.copyOfRange(0, cut), data.copyOfRange(cut, data.size))
                }
                ExtendedFragmentStrategy.FULL5 -> fixedChunks(data, 5)
                ExtendedFragmentStrategy.FULL10 -> fixedChunks(data, 10)
                ExtendedFragmentStrategy.FULL20 -> fixedChunks(data, 20)
                ExtendedFragmentStrategy.MULTI64 -> fixedChunks(data, if (chunkSize > 0) chunkSize else 64)
                ExtendedFragmentStrategy.SNI_BOUNDARY -> {
                    val cut = if (location != null) location.first.coerceIn(1, data.size - 1) else data.size / 2
                    listOf(data.copyOfRange(0, cut), data.copyOfRange(cut, data.size))
                }
                ExtendedFragmentStrategy.SNI_SPLIT -> {
                    val cut = if (location != null) {
                        (location.first + maxOf(1, location.second / 2)).coerceIn(1, data.size - 1)
                    } else {
                        data.size / 2
                    }
                    listOf(data.copyOfRange(0, cut), data.copyOfRange(cut, data.size))
                }
                ExtendedFragmentStrategy.SNI_CHARS -> {
                    if (location != null) {
                        val out = mutableListOf<ByteArray>()
                        if (location.first > 0) {
                            out.add(data.copyOfRange(0, location.first))
                        }
                        for (i in 0 until location.second) {
                            out.add(byteArrayOf(data[location.first + i]))
                        }
                        if (location.first + location.second < data.size) {
                            out.add(data.copyOfRange(location.first + location.second, data.size))
                        }
                        out
                    } else {
                        val cut = maxOf(1, data.size / 2)
                        listOf(data.copyOfRange(0, cut), data.copyOfRange(cut, data.size))
                    }
                }
                ExtendedFragmentStrategy.TLS_RECORD_FRAG -> {
                    splitTlsRecord(data, atSni = false, location = location)
                }
                ExtendedFragmentStrategy.TLS_SNI_RECORDS -> {
                    splitTlsRecord(data, atSni = true, location = location)
                }
            }
        }

        private fun fixedChunks(data: ByteArray, size: Int): List<ByteArray> {
            val safeSize = maxOf(1, size)
            return (data.indices step safeSize).map { offset ->
                data.copyOfRange(offset, minOf(offset + safeSize, data.size))
            }
        }

        private fun splitTlsRecord(data: ByteArray, atSni: Boolean, location: Pair<Int, Int>?): List<ByteArray> {
            if (data.size < 6 || (data[0].toInt() and 0xff) != 0x16) return listOf(data.copyOf())
            val payload = data.copyOfRange(5, data.size)
            if (payload.size < 2) return listOf(data.copyOf())
            val split = if (atSni && location != null) {
                (location.first - 5).coerceIn(1, payload.size - 1)
            } else {
                maxOf(1, payload.size / 2)
            }
            val version = data.copyOfRange(1, 3)
            return listOf(
                tlsRecord(version, payload.copyOfRange(0, split)),
                tlsRecord(version, payload.copyOfRange(split, payload.size)),
            )
        }

        private fun tlsRecord(version: ByteArray, payload: ByteArray): ByteArray = byteArrayOf(
            0x16,
            version[0],
            version[1],
            ((payload.size ushr 8) and 0xff).toByte(),
            (payload.size and 0xff).toByte(),
        ) + payload

        fun buildDisposableFakeProbe(fakeSni: String): ByteArray {
            val hostBytes = fakeSni.toByteArray(StandardCharsets.US_ASCII)
            val pkt = mutableListOf<Byte>()

            pkt.addAll(listOf(0x16.toByte(), 0x03.toByte(), 0x01.toByte(), 0x00.toByte(), 0x00.toByte()))
            val handshakeStart = pkt.size
            pkt.addAll(listOf(0x01.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte()))
            pkt.addAll(listOf(0x03.toByte(), 0x03.toByte()))

            for (i in 0 until 32) {
                pkt.add(((i * 37 + 11) and 0xff).toByte())
            }
            pkt.add(0x00.toByte())

            pkt.addAll(listOf(0x00.toByte(), 0x04.toByte(), 0x13.toByte(), 0x01.toByte(), 0x13.toByte(), 0x03.toByte()))
            pkt.addAll(listOf(0x01.toByte(), 0x00.toByte()))

            val extLenPos = pkt.size
            pkt.addAll(listOf(0x00.toByte(), 0x00.toByte()))
            val extStart = pkt.size

            val sniExtLen = 2 + 1 + 2 + hostBytes.size
            pkt.addAll(listOf(0x00.toByte(), 0x00.toByte(), ((sniExtLen ushr 8) and 0xff).toByte(), (sniExtLen and 0xff).toByte()))
            val listLen = 1 + 2 + hostBytes.size
            pkt.addAll(listOf(((listLen ushr 8) and 0xff).toByte(), (listLen and 0xff).toByte(), 0x00.toByte(), ((hostBytes.size ushr 8) and 0xff).toByte(), (hostBytes.size and 0xff).toByte()))
            for (b in hostBytes) pkt.add(b)

            pkt.addAll(listOf(0x00.toByte(), 0x10.toByte(), 0x00.toByte(), 0x0e.toByte(), 0x00.toByte(), 0x0c.toByte(), 0x08.toByte(), 'h'.code.toByte(), 't'.code.toByte(), 't'.code.toByte(), 'p'.code.toByte(), '/'.code.toByte(), '1'.code.toByte(), '.'.code.toByte(), '1'.code.toByte(), 0x02.toByte(), 'h'.code.toByte(), '2'.code.toByte()))

            val extLen = pkt.size - extStart
            pkt[extLenPos] = ((extLen ushr 8) and 0xff).toByte()
            pkt[extLenPos + 1] = (extLen and 0xff).toByte()

            val hsLen = pkt.size - handshakeStart - 4
            pkt[handshakeStart + 1] = ((hsLen ushr 16) and 0xff).toByte()
            pkt[handshakeStart + 2] = ((hsLen ushr 8) and 0xff).toByte()
            pkt[handshakeStart + 3] = (hsLen and 0xff).toByte()

            val recLen = pkt.size - 5
            pkt[3] = ((recLen ushr 8) and 0xff).toByte()
            pkt[4] = (recLen and 0xff).toByte()

            return pkt.toByteArray()
        }
    }
}
